### ISSUE TITLE
Incorrect @BeforeClass order in IntelliJ

### ISSUE THREAD

**Issue by pauljct06** [junit-team/junit4#1372]
Here is my Test class
![image](https://cloud.githubusercontent.com/assets/9300208/19398206/80a02bde-9219-11e6-8122-8cf22090389e.png)

@BeforeClass is running at the end when selecting run
<img width="849" alt="screen shot 2016-10-14 at 2 16 43 pm" src="https://cloud.githubusercontent.com/assets/9300208/19398132/1cd029a6-9219-11e6-8a34-a328423e193c.png">

but when running the test as a debug it comes in the right order
![image](https://cloud.githubusercontent.com/assets/9300208/19398280/e41a84c0-9219-11e6-8c4b-db5abad209fd.png)

**Comment by BrunoDM2943** [comment]
Wich junit's version are you using? Have you tried run this on other IDE?

**Comment by kcooney** [comment]
I suspect IntelliJ might be capturing stout and printing it to the console later. You can use a debugger to see what order the methods are called.

**Comment by pauljct06** [comment]
I am using IntelliJ Ultimate 2016.2.4. I haven't had the time to try it on another IDE yet.

**Comment by marcphilipp** [comment]
We had a similar problem with IntelliJ in JUnit 5: https://youtrack.jetbrains.com/issue/IDEA-160397

**Comment by BrunoDM2943** [comment]
Guys, I used hamcrest-all-1.3.jar, junit-4.12 and Intellij Ultimate 2016.2.4. This bug didnt happened! @pauljct06 , could you attach the junit and hamcrest jars that you are using? 

![image](https://cloud.githubusercontent.com/assets/7273952/19451087/bcbadaf6-948a-11e6-959e-3cbcb72dd2dc.png)

**Comment by pauljct06** [comment]
@BrunoDM2943  Sorry for the delay, see below the Junit and the Hamcrest I am using

![image](https://cloud.githubusercontent.com/assets/9300208/19493151/8d82dbae-9547-11e6-8bd7-5f868caa53bc.png)

**Comment by kcooney** [comment]
@pauljct06 I think IntelliJ is buffering the output, which could affect the output order. Please use a debugger to see what order the methods are being called. Alternatively, modify your test to have a private static List field and update your test methods to append the name of the list to the field and then assert at the end that you get the expected order.

**Comment by marcphilipp** [comment]
@pauljct06 Which version of IDEA are you using? Can you update and try again?

**Comment by davidonlaptop** [comment]
@kcooney : It does look like IDEA is buffering the output:

```java
public class IntellijJunitTest
{
    private static boolean setupCompleted = false;

    @BeforeClass
    public static void setUp() {
        System.out.println("@BeforeClass");
        setupCompleted = true;
    }

    @AfterClass
    public static void tearDown() {
        System.out.println("@AfterClass: " + setupCompleted);
    }

    @Before
    public void before() {
        System.out.println("@Before: " + setupCompleted);
    }

    @After
    public void after() {
        System.out.println("@After: " + setupCompleted);
    }

    @Test
    public void test1() {
        System.out.println("*** test1: " + setupCompleted);
    }

    @Test
    public void test2() {
        System.out.println("*** test2: " + setupCompleted);
    }

    @Test
    public void test3() {
        System.out.println("*** test3: " + setupCompleted);
    }

    @Test
    public void test4() {
        System.out.println("*** test4: " + setupCompleted);
    }

    @Test
    public void test5() {
        System.out.println("*** test5: " + setupCompleted);
    }
}
```

Output:
```
@Before: true
*** test1: true
@After: true
@Before: true
*** test2: true
@After: true
@Before: true
*** test3: true
@After: true
@Before: true
*** test4: true
@After: true
@Before: true
*** test5: true
@After: true
@BeforeClass
@AfterClass: true
```

**Comment by kcooney** [comment]
@davidonlaptop thank! So not a bug on our side :-)

**Comment by davidonlaptop** [comment]
For reference: this bug appears to be fixed in the latest IntelliJ version (2016.3.1 RC)
https://intellij-support.jetbrains.com/hc/en-us/requests/825635?page=1
